src/lib/o1trace/o1trace.ml

```diff
@@ -0,0 +1,116 @@
+open Core
+open Async
+
+[%%import
+"../../config.mlh"]
+
+[%%if
+tracing]
+
+module Scheduler = Async_kernel_scheduler
+
+type event_kind =
+  | New_thread
+  | Thread_switch
+  | Cycle_start
+  | Cycle_end
+  | Pid_is
+  | Event
+  | Start
+  | End
+
+let buf = Bigstring.create 128
+
+let emitk (k : event_kind) pos =
+  let num =
+    match k with
+    | New_thread -> 0
+    | Thread_switch -> 1
+    | Cycle_start -> 2
+    | Cycle_end -> 3
+    | Pid_is -> 4
+    | Event -> 5
+    | Start -> 6
+    | End -> 7
+  in
+  Bigstring.set_uint8 buf ~pos num ;
+  pos + 1
+
+let emiti (i : int) pos =
+  Bigstring.set_uint64_le buf ~pos i ;
+  pos + 8
+
+let emits (s : string) pos =
+  let sl = String.length s in
+  let pos = emiti sl pos in
+  Bigstring.From_string.blit ~src:s ~src_pos:0 ~len:sl ~dst:buf ~dst_pos:pos ;
+  pos + sl
+
+let emitt pos =
+  let ts = Time_stamp_counter.to_time_ns (Time_stamp_counter.now ()) in
+  emiti (Core.Time_ns.to_int63_ns_since_epoch ts |> Int63.to_int_exn) pos
+
+let finish wr final_len = Writer.write_bigstring wr ~pos:0 ~len:final_len buf
+
+let trace_new_thread' wr (name : string) (ctx : Execution_context.t) =
+  emitk New_thread 0 |> emitt |> emiti ctx.tid |> emits name |> finish wr
+
+let trace_thread_switch' wr (new_ctx : Execution_context.t) =
+  emitk Thread_switch 0 |> emitt |> emiti new_ctx.tid |> finish wr
+
+let tid = ref 1
+
+let trace_event' wr (name : string) =
+  emitk Event 0 |> emitt |> emits name |> finish wr
+
+let trace_event_impl = ref (fun _ -> ())
+
+let trace_event name = !trace_event_impl name
+
+let trace_task (name : string) (f : unit -> 'a) =
+  let new_ctx =
+    Execution_context.with_tid
+      (Scheduler.current_execution_context (Scheduler.t ()))
+      !tid
+  in
+  tid := !tid + 1 ;
+  !Async_kernel.Tracing.fns.trace_new_thread name new_ctx ;
+  match Scheduler.within_context new_ctx f with
+  | Error () ->
+      failwith "traced task failed, exception reported to parent monitor"
+  | Ok x -> x
+
+let measure' wr (name : string) (f : unit -> 'a) : 'a =
+  emitk Start 0 |> emitt |> emits name |> finish wr ;
+  let res = f () in
+  emitk End 0 |> emitt |> finish wr ;
+  res
+
+let measure_impl = ref (fun _ f -> f ())
+
+(* the two things we set measure_impl to are fine *)
+
+let measure name (f : unit -> 'a) : 'a =
+  (Obj.magic !measure_impl : string -> (unit -> 'a) -> 'a) name f
```

Comment:
do you really need an obj.magic here? Did you try putting a type annotation on the measure_impl from 89?

Reply:
yes, value restriction. can't stick an `(string -> () -> 'a) -> 'a` in a ref, `'a` becomes weak.

Reply:
Even if you use explicit forall?

Reply:
yes, you try it: `let measure_impl : type a. (string -> (unit -> a) -> a) ref = ref (fun _ f -> f ())`. https://caml.inria.fr/pub/papers/garrigue-value_restriction-fiwflp04.pdf may enlighten you.

Reply:
(functions aren't values, so when you stick them in a mutable location they lose their polymorphism)